### PR TITLE
instruction-level generic join

### PR DESCRIPTION
**Pull request by havardpe**
also move some stuff out of simple_value

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@arnej27959 please review

### CODE REVIEW

File: eval/src/vespa/eval/eval/tensor_instructions.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "interpreted_function.h"
+
+namespace vespalib {
+
+class Stash;
+
+namespace eval {
+
+struct JoinPlan;
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/tensor_instructions.h:13]: There is no "JoinPlan".

File: eval/src/vespa/eval/eval/tensor_instructions.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace tensor_instruction {
+
+using join_fun_t = double (*)(double, double);
+
+InterpretedFunction::Instruction make_join(const ValueType &lhs_type, const ValueType &rhs_type, join_fun_t function,
+                                           const ValueBuilderFactory &factory, Stash &stash);
+
+}
+}
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/tensor_instructions.h:23]: ending three nested namespaces like this feels confusing to me.
I suggest something like:

```
namespace vespalib { class Stash; }

namespace vespalib::eval::tensor_instruction {
... rest of file ...
}
```

File: eval/src/vespa/eval/eval/tensor_plans.h [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+struct DenseJoinPlan {
+    size_t lhs_size;
+    size_t rhs_size;
+    size_t out_size;
+    std::vector<size_t> loop_cnt;
+    std::vector<size_t> lhs_stride;
+    std::vector<size_t> rhs_stride;
+    DenseJoinPlan(const ValueType &lhs_type, const ValueType &rhs_type);
+    ~DenseJoinPlan();
+    template <typename F> void execute(size_t lhs, size_t rhs, F &&f) const {
```
Comment (arnej27959) [inline eval/src/vespa/eval/eval/tensor_plans.h:30]: is it possible to move the various execute() implementations to the .cpp file? or would that lead to less inlining opportunities?

**Review by arnej27959 — APPROVED** [review]
👍 only cosmetic comments 👍

File: eval/src/vespa/eval/eval/tensor_plans.h (hunk 3 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/tensor_plans.h:30]: I think we want it in a separate header, for cross-use in different plans. We do not want to call the function via pointer, so I think we should avoid having it in a cpp file.

File: eval/src/vespa/eval/eval/tensor_instructions.h (hunk 1 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/tensor_instructions.h:13]: it got renamed to JoinParam and made local, I will clean it up

File: eval/src/vespa/eval/eval/tensor_instructions.h (hunk 2 above)
Comment (havardpe) [reply eval/src/vespa/eval/eval/tensor_instructions.h:23]: agree

**Pull request merged**